Studio-Lovelies/GG-JointJustice-Unity
# Added save file management (via C# and UNLOCK_CHAPTER action line)

## Summary
This PR implements a persistent storage using Unity's [`PlayerPrefs`](https://docs.unity3d.com/ScriptReference/PlayerPrefs.html) and the classes inside the `Savefiles` namespace.

---

#### [Savefiles.SaveData](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/blob/feature/116-savefiles/unity-ggjj/Assets/Scripts/Savefiles/SaveData.cs)
A `plain-old data object`/`passive object`/`pure data structure` which data is serialized.  
Constructing an instance of it by hand, results in whatever we want the default settings to be.  
Currently only contains a version number and a [bit field](https://docs.microsoft.com/en-us/dotnet/api/system.flagsattribute?view=netframework-4.8) of unlocked chapters for testing purposes.

#### [Savefiles.Proxy](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/blob/feature/116-savefiles/unity-ggjj/Assets/Scripts/Savefiles/Proxy.cs)
This class is our layer of abstraction to Unity's persistent storage:
- `HasExistingSaveData()` returns a bool
- `Load()` throws if the method above returns false or returns the `Savefiles.SaveData` currently stored
- `UpdateCurrentSaveData(callback)` takes a function that can modify save data, runs it and then saves the values (a good example is the [`UNLOCK_CHAPTER`-action implementation](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/blob/feature/116-savefiles/unity-ggjj/Assets/Scripts/TextDecoder/ActionDecoder.cs#L619-L628))

This class also takes care of upgrading save files between game versions. `UpgradeSaveData()` is currently basically empty and only needs to change, if we ever make breaking changes between versions. (For example if we first store the screen resolution in two `float`s but later decide to move it to a `Vector2`.)



## Before/after screenshots and/or animated gif

### Unrelated, minor change: 
#### Made the error message you get when using a new Enum of Class inside `ActionDecoder` which has no test data a bit more readable:
##### Before
![image](https://user-images.githubusercontent.com/1689033/150484815-59e0ad85-ccdc-4ae4-a55b-caffd23f5583.png)
##### After
![image](https://user-images.githubusercontent.com/1689033/150484830-def58066-ecac-4f6c-b5d3-d083970b849c.png)

## Testing instructions

1. Play the `Savefiles - TestScene.unity` once to unlock chapters
2. Play the scene again, notice chapters already being unlocked
3. Without running the game, inside the Unity Editor, press `Cmd+Shift+D`/`Ctrl+Shift+D` or go to `Edit/Delete Save Data`
4. Play the `Savefiles - TestScene.unity` scene again and notice all previously unlocked chapters being locked
5. Make sure existing and the [newly added tests](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/blob/feature/116-savefiles/unity-ggjj/Assets/Tests/EditModeTests/SaveDataTests.cs) complete successfully


## Additional information

- `[ ]` Changes UI
- `[x]` Introduces new feature
  - Savefiles
- `[ ]` Removes existing feature
- `[x]` Has associated resource: 
  - Closes https://github.com/Studio-Lovelies/GG-JointJustice-Unity/issues/116
